MAINT: remove exec_command usage from ccompiler.py

tylerjereddy opened the pull request:
Related to #11824.

An interesting side effect of this branch -- NumPy builds actually properly use my terminal color scheme on mac--so warnings and underlines from compiler are properly highlighted (just all-grey on master).

A few other notes here:

- there's partial coverage by tests according to codecov
- had to reconstitute some of the string processing from exec_command after check_output; NumPy fails to build otherwise

tylerjereddy commented:
The NumPy build in CircleCI has been failing here because `OSError` and `CalledProcessError` have different attributes / hooks for grabbing output / error codes; I've initially revised the apparently problematic case, but the other usage in this PR only has a `NOTE` added for now since that may not cause an immediate problem here, but should likely be fixed.

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +625,20 @@ def matcher(version_string):
-    status, output = exec_command(version_cmd, use_tee=0)
+    try:
+        output = subprocess.check_output(version_cmd)
+        output = output.decode(locale.getpreferredencoding(False),
```
This stuff will probably be abstracted to a helper function as [Eric suggested in a related PR](https://github.com/numpy/numpy/pull/11900#discussion_r216027013), and probably moved to an `else` block as done there too.

tylerjereddy commented:
Revised to reflect #11900, and while tests pass locally, something really nasty happens rather silently--way more tests are skipped:

`SKIP [396] /Users/treddy/miniconda3/envs/numpy_dev_py36/lib/python3.6/site-packages/_pytest/nose.py:23: No C compiler available`

Obviously, that is scary because it can look ok depending on how closely one inspects the full test suite output.

I'm going to add a WIP tag here for safety while I investigate this--obviously messing with the C compiler, even in a subtle way, can be bad for NumPy if not done properly.

charris commented:
Why is `_pytest/nose.py` showing up? We should have nothing to do with nose and I don't think anything we access should require nose unless there is some leftover bit that didn't get properly updated on the change. There are nose dependent tests in `numpy/testing/tests`, but they are only needed to test nose dependent decorators for downstream back-compatibility and should be skipped if there is no nose.

charris commented:
Thanks for digging through this, numpy distutils is one of the oldest and least comprehensible parts of NumPy.

charris commented:
Looks like a missing return value somewhere.

tylerjereddy commented:
Now works locally in the version I just pushed-- really have to be careful to not trust the docs about the type (`str` vs. `list`) of input fed to `subprocess` from various utility functions.

Will leave WIP for now, another set of eyes should comb through even if the test matrix is green with constant skip counts, especially for the C compiler infrastructure.

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,54 @@ def CCompiler_spawn(self, cmd, display=None):
+        # though
+        s = 127
+    else:
+        o = filepath_from_subprocess_output(o)
```
this line is pointless, I think

eric-wieser replied:
this isn't reasonable to call unless the goal is actually to get a filename from the thing that was spawned

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,54 @@ def CCompiler_spawn(self, cmd, display=None):
+        # use a convenience return here so that any kind of
+        # caught exception will execute the default code after the
+        # try / except block, which handles various exceptions
+        return None
```
feels slightly non-Pythonic to use a convenience return maybe?

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,54 @@ def CCompiler_spawn(self, cmd, display=None):
+    except subprocess.CalledProcessError as exc:
+        o = exc.output
+        s = exc.returncode
+    except (OSError, AttributeError):
```
I think I can remove the `AttributeError` catch now that  `split()` is guarded by the `is_sequence` check. The latter `is_sequence` check seems like something the standard library may also have something suitable for, though I should perhaps confine refactoring to essential stuff

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +655,26 @@ def matcher(version_string):
+    except subprocess.CalledProcessError as exc:
+        output = exc.output
+        status = exc.returncode
+    except (OSError, AttributeError):
```
`AttributeError` can probably be extracted now that `is_sequence` guards `version_cmd` here too

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,54 @@ def CCompiler_spawn(self, cmd, display=None):
+        # or multiple commands; historically, exec_command() could
+        # handle string input directly
+        if not is_sequence(cmd):
+            cmd = cmd.split(' ')
```
this will be wrong if any arguments contain spaces. I think your best bet is just to pass the string on unmodified, and let subprocess handle it.

eric-wieser replied:
(and in one other place too)

eric-wieser reviewed (changes requested):
Splitting commands on spaces is not safe. Better to just pass them on to subprocess and let it do its best to split them.

charris commented:
@eric-wieser OK with you?

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,51 @@ def CCompiler_spawn(self, cmd, display=None):
+        # When installing through pip, `o` can contain non-ascii chars
+        pass
+    # NOTE: the str() operation here appears necessary on Python 3.x
+    if re.search('Too many open files', str(o)):
```
Changing the search string to `b'Too many ...'`  would eliminate the need for the `str`

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,51 @@ def CCompiler_spawn(self, cmd, display=None):
+        # check_output wants a list -- we use split() to
+        # produce a list irrespective of whether cmd is a single
+        # or multiple commands; historically, exec_command() could
+        # handle string input directly
```
This comment looks out of date

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,51 @@ def CCompiler_spawn(self, cmd, display=None):
+        # output, but exec_command() historically would use an
+        # empty string for EnvironmentError (base class for
+        # OSError)
+        # NOTE: catching AttributeError is new after migrating
```
This comment is outdated too

eric-wieser reviewed (dismissed):
@charris: There look to be some comments left from an intermediate revision of this patch that either don't belong, or are missing the code that goes with them

tylerjereddy commented:
Yeah, my notes concur that I still need clean-up here--I was / am focused on getting the f2py one satisfactory first--which has proven non-trivial.

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,39 @@ def CCompiler_spawn(self, cmd, display=None):
-            msg = ''
-        raise DistutilsExecError('Command "%s" failed with exit status %d%s' % (cmd, s, msg))
+    try:
+        o = subprocess.check_output(cmd)
```
`cmd` is documented to be a string, but in practice tends to be a list it seems; I originally tried to split on whitespace just in case, which Eric rightly asked me to remove -- I wonder if we'd consider adding a conditional  `shlex.split()` here for robustness given other recent similar changes.

tylerjereddy replied:
Also, the LGTM analysis wants me to remove assignment to `o` here, since it is never used.

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +640,19 @@ def matcher(version_string):
 
-    status, output = exec_command(version_cmd, use_tee=0)
+    try:
+        output = subprocess.check_output(version_cmd)
```
similar comment re: string vs. list robustness

tylerjereddy commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +640,19 @@ def matcher(version_string):
+        status = 127
+        output = ''
+    else:
+        output = filepath_from_subprocess_output(output)
```
I think we're expecting a version string here and Eric has previously criticized extraneous usage of our new `filepath_from_subprocess_output` -- it may be possible to remove this, but note that `exec_command` was not able to make the distinction between filepaths and version strings so it would also perform the processing steps / control flow invoked by this function anyway.

tylerjereddy reviewed (commented):
I revised to remove some extraneous comments and the str fix; added a few comments on the tricky parts.

tylerjereddy commented:
It looks to me like one of the requested revisions is causing C compiler detection failure -- that actually passes through most of the CI (notice huge skip counts in travis coverage job though), but the coverage does rightly indicate a problem with the large number of skipped tests with total project coverage dropping almost 4 %.

I've seen this before on this branch.

The latest revision was all comment-related changes except for swapping `str` coercion for a b-prefixed bytes string.

I note that this only happens for a subset of the travis / linux jobs, and the Windows jobs I checked looked ok. Large numbers of new skips can be silent failures if we don't inspect manually or track coverage on the affected jobs.

charris commented:
Might be worth trying a build with tcsh after all this is in.

tylerjereddy commented:
Revised to restore string coercion on the regular expression search -- checking to see if this alleviates the issue of hundreds of additional tests being skipped in a subset of the CI jobs.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,39 @@ def CCompiler_spawn(self, cmd, display=None):
+        # output, but exec_command() historically would use an
+        # empty string for EnvironmentError (base class for
+        # OSError)
+        o = ''
```
This should be `b''` to match the type of the success case. With that change, the `str` below can become a bytes prefix again

tylerjereddy replied:
is `exc.output` above guaranteed to be bytes?

eric-wieser replied:
Yes, from the docs: _"By default, this function will return the data as encoded bytes."_

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +640,19 @@ def matcher(version_string):
+        # match the historical returns for a parent
+        # exception class caught by exec_command()
+        status = 127
+        output = ''
```
b prefix

tylerjereddy commented:
Revised based on feedback from Eric--we'll see what the CI says. May also be helpful to try building i.e., SciPy master branch with this branch locally as a sort of in-the-wild test for distutils mods, as Stefan / Matti suggested.

tylerjereddy commented:
The subtle Travis CI failure for the Python 3.6 coverage job is perhaps related to the environment variable modifications that result in `gfortran --coverage ` being used in place of plain `gfortran` (and so on for compilation of C as well). I'll see if I can reproduce locally with those env vars--I certainly see the `--coverage` flag on the traceback in Travis. Maybe we'll indeed have to add a `shlex.split()` here too as alluded to above.

charris commented:
Not sure why this is suddenly showing up in `test_quoted_character.py`.
```
                 from /tmp/tmppfaghj0k/src.linux-x86_64-3.6/fortranobject.h:13,
                 from /tmp/tmppfaghj0k/src.linux-x86_64-3.6/_test_ext_module_5405module.c:15:
/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
```

tylerjereddy commented:
@charris That test is almost brand new (recent merge) -- what is surprising about the message?

It was clear that the `cmd` variable here can be a list **or** a string (the docs are dead wrong -- should probably fix that too), so I've pushed something that looks like our usual `shlex` handling on this set of PRs.

charris commented:
I was surprised that none of the other tests show that error. I don't see it  locally with this PR, so I'm probably missing something.

charris commented:
BTW, when making these tests you can add `[skip appveyor]` to the first commit message line after heading so that appveyor tests are not run.

charris commented:
I note that the `--debug` flag is used in that test module, maybe that is related?

tylerjereddy commented:
I fixed the test failure with a crude patch related to F77 compiler environment variable settings (i.e, `F77=gfortran --coverage` and so on seems to be able to mess with our internal executable detection machinery, which perhaps mostly assumes we use things like `extra_args` to feed flags in later).

We still have more subtle issues here though -- the bulk of spurious test skips are from:
`SKIP [23] /home/treddy/miniconda2/envs/numpy_dev_py3/lib/python3.6/site-packages/_pytest/nose.py:24: No Fortran 90 compiler available`

That traces to `pytest_runtest_makereport()` function in `nose.py`, which seems to handle skips in some cases. To replicate the CI job I use the coverage compiler environment variable settings locally, so perhaps not surprising that f90 executable detection also needs to compensate for that somehow, though not sure why it needs different handling vs. f77.

tylerjereddy commented on numpy/distutils/fcompiler/__init__.py:
```diff
@@ -465,8 +466,11 @@ def customize(self, dist = None):
-        f77 = self.command_vars.compiler_f77
-        f90 = self.command_vars.compiler_f90
+        # TODO: proper guards here and same for F90?
+        f77 = shlex.split(self.command_vars.compiler_f77,
```
This seems to help from brute force debugging on the single test failure, but probably isn't right--we may for example just lose the `--coverage` flag completely by doing this instead of putting it in `extra_args` somewhere else.

I guess I'll have to drill down on that.

eric-wieser replied:
Can you drop the `[0]`, and use `(compiler_fix=fix+fixflags+fflags)` instead of `(compiler_fix=[fix]+fixflags+fflags)` below?

tylerjereddy replied:
I tried it just now, but it breaks NumPy build:

```
  File "/Users/treddy/github_projects/numpy/numpy/distutils/fcompiler/__init__.py", line 524, in customize
    self.set_commands(compiler_fix=fix+fixflags+fflags)
TypeError: must be str, not list
```

eric-wieser replied:
That's probably because you didn't drop the `[0]` on line 493

tylerjereddy replied:
That's not quite enough--the types of `f77` and `f90` variables have changed with indexing removal so I needed a few more edits locally.

charris commented:
Any idea on why nose is involved? Is there some leftover nose construct that is calling through a compatibility layer?

charris commented:
I wonder if there is a way to turn off nose compatibility?

tylerjereddy commented:
It seems to be intercepting on `SkipTest`
- `numpy/f2py/tests/util.py:            raise SkipTest("No Fortran 90 compiler available")`
- `numpy/core/tests/_locales.py:            raise SkipTest("No French locale available")`

charris commented:
I checked `SkipTest`, it comes from unittest, not nose, but maybe it just gets tossed in there.

tylerjereddy commented:
Looks like `pytest` doesn't care (!) if you use `unittest` or nose -- if you don't use their own skipping mechanism it will get routed through the `_pytest/nose.py `

from `get_skip_exceptions()` in above pytest module:

```Python
for module_name in ('unittest', 'unittest2', 'nose'):
    mod = sys.modules.get(module_name)
    if hasattr(mod, 'SkipTest'):
...
```

To be fair, `nose` also has `SkipTest` so that may have been the most straightforward way for them to handle it.

tylerjereddy commented:
Should I perhaps just open an issue summarizing that and return my focus to the matter at hand (fortran compiler command stuff) for now? I think the problem just gets routed through the compatibility layer but the compatibility layer is just relaying the underlying issue.

charris commented:
Yep, that would be helpful. Looks like the way to dynamically skip a test is to call `pytest.skip`.

tylerjereddy commented:
I was able to rectify the issue with the remaining compiler type locally by splitting on the `fix` compiler; pretty sure this just excludes any extra flags (like `--coverage`). We could probably put those in a compiler arguments variable somewhere so they aren't lost, but then should also add tests that probe that the env-variable specified flags are retained in addition to `extra_args`.

Let's see if the CI agrees that extra tests aren't being skipped (i.e., project coverage should not drop substantially) as a start.

tylerjereddy commented:
CI results looking better now -- project coverage increases slightly, and --coverage job skip counts are back down to normal levels it seems.

I did a local (WSL / Ubuntu / Python 3.6 / bash) test for building latest SciPy master (fdd27a3) using clean builds of NumPy master (6f0a0faf), vs. the tip of this PR branch. In both cases, the clean SciPy builds produced identical full test suite results with: `python runtests.py --mode=full -- -n 3 -rsx`

`14904 passed, 1462 skipped, 112 xfailed, 9 xpassed`

One thing I did notice in the two SciPy builds--the newer distutils build allows the compile warnings to be shown in color, while the NumPy master distutils is all grey on my machine (maybe an indication of less interference with local shell settings). I also had the `--coverage` compiler flags on for all NumPy & SciPy builds in this test scenario, just to make things more likely to fail.

I guess the next thing is coming up with a suitable / sensible tcsh-based test.

charris commented:
Don't know about tcsh
```
charris@fc [numpy.git ((v1.15.1))]$ tcsh -c "python runtests.py -mfull"
...
charris@fc [numpy.git ((v1.15.1))]$ tcsh --version
tcsh 6.20.00 (Astron) 2016-11-24 (x86_64-unknown-linux) options wide,nls,dl,al,kan,sm,rh,color,filec
```
Works fine. The problem may be specific to the os, local environment, and problem. I say we go ahead and commit, then ask the folks who had problems to see if they are fixed.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +140,53 @@ def CCompiler_spawn(self, cmd, display=None):
         else:
-            msg = ''
-        raise DistutilsExecError('Command "%s" failed with exit status %d%s' % (cmd, s, msg))
+            subprocess.check_output(cmd)
```
Why can't we let `subprocess` do the splitting internally?

tylerjereddy replied:
I thought I had a good reason for doing that, but maybe not now that I've hacked things downstream in the control flow. Things seem to work locally if I remove the splitting block above, even with coverage flags exported in env.

I'm slightly nervous to make more changes now because of the extreme sensitivity to changes in this PR branch, but perhaps I could remove that split & squash / push.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,39 @@ def CCompiler_spawn(self, cmd, display=None):
+    except UnicodeError:
+        # When installing through pip, `o` can contain non-ascii chars
+        pass
+    if re.search(b'Too many open files', o):
```
Both places above are now of type `bytes`, so yes, this is safe.

tylerjereddy commented:
Revised based on Eric's most recent comment, rebased, squashed, and force pushed.

If the CI is all green I suppose we can remove the WIP tag. I note that the handling of environment variables that modify compiler exe settings (i.e., `F77 = gfortran --coverage`) is probably sub-optimal since I think this branch will just drop those, but not clear that any robust handling was in place for this before either.

tylerjereddy committed:
MAINT: remove exec_command usage from ccompiler.py

* replaced the usage of exec_command() in
ccompiler distutils module with subprocess
equivalent

tylerjereddy commented:
Revised to address extraneous import flagged by LGTM analysis.

eric-wieser commented:
Taking a step back here - why does this function exist at all? What's broken with the default CCompiler.spawn?

mattip commented:
One difference is the integration `with log.info(display)`, where `display` is a kwarg that can be different from `cmd`

charris commented:
@eric-wieser How does this look to you now? Note that the immediate problem is to remove shell dependence, further cleanups would be nice, but are supplementary.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -620,7 +640,19 @@ def matcher(version_string):
+        status = 127
+        output = b''
+    else:
+        output = filepath_from_subprocess_output(output)
```
Strictly speaking this is not a filepath, but you're correct that this matches the old behavior. Add a comment to that effect, and then this looks good to me.

eric-wieser commented on numpy/distutils/fcompiler/__init__.py:
```diff
@@ -465,8 +466,11 @@ def customize(self, dist = None):
 
-        f77 = self.command_vars.compiler_f77
-        f90 = self.command_vars.compiler_f90
+        # TODO: proper guards here and same for F90?
```
Not sure what you mean by this comment

tylerjereddy commented:
Revised again to remove the various `[0]` indices as requested by Eric

tylerjereddy commented on numpy/distutils/fcompiler/__init__.py:
```diff
@@ -480,6 +483,14 @@ def customize(self, dist = None):
+        # this and similar examples where a Fortran compiler
+        # environment variable has been customized by CI or a user
+        # should perhaps eventually be more throughly tested and more
+        # robustly handled
```
I suppose this comment may be out of date after Eric's `[0]` removal suggestion, though not confident compiler flag handling is perfect at the moment.

rgommers commented:
Would be good to merge this soon, to give it a bit of time in master before 1.16.x branching. Looks like it's in good shape to me.

@eric-wieser I resolved all comments you make that have been addressed. Could you have another look?

eric-wieser reviewed (commented):
Regarding splitting the compiler names, the old behavior was (checks indicate the intent implied by the variable contents being followed):

|F77 env var|window|posix|
|--|--|---|
|`f77 --coverage`| :x: | :heavy_check_mark: |
|`path/with spaces/f77`| :heavy_check_mark: | :x: |

And the new behavior is now

|F77 env var|window|posix|
|--|--|---|
|`f77 --coverage`| :heavy_check_mark: | :heavy_check_mark: |
|`path/with spaces/f77`| :x: | :x: |

It's not clear to me whether this is a better or worse choice than

|F77 env var|window|posix|
|--|--|---|
|`f77 --coverage`| :x: | :x: |
|`path/with spaces/f77`| :heavy_check_mark: | :heavy_check_mark: |

Which would be acheived by replacing `f77 = shlex.split` with `f77 = [f77]`

---

Alternatively,  we could keep the old platform-dependent behavior, and leave fixing it for another PR, which would look like:
```
if os.platform == 'posix':
    fix = shlex.split(f77)
else:
    fix = [f77]
```

tylerjereddy commented:
@eric-wieser 

The following seems to help deal with both F77 pass-throughs locally, but we've previously had discussions about the dangers of custom shell parsing in this effort so not sure what your view on this is--the corner cases seem at least rare?

```diff
--- a/numpy/distutils/fcompiler/__init__.py
+++ b/numpy/distutils/fcompiler/__init__.py
@@ -466,8 +466,15 @@ class FCompiler(CCompiler):
         noarch = self.distutils_vars.get('noarch', noopt)
         debug = self.distutils_vars.get('debug', False)
 
-        f77 = shlex.split(self.command_vars.compiler_f77,
-                          posix=(os.name == 'posix'))
+        # if the F77 env variable contains compiler flags
+        # (should start with - or --), splitting on the space
+        # that precedes the first dash allows preservation
+        # of whitespace in F77 base command / path
+        # NOTE: works as long as there's no consecutive space + dash in
+        # path, which seems pretty pathological anyway
+        f77 = self.command_vars.compiler_f77.split(' -')[0]
+        f77_env_flags = self.command_vars.compiler_f77[len(f77):]
+        f77_env_flags = shlex.split(f77_env_flags, posix=(os.name == 'posix'))
         f90 = shlex.split(self.command_vars.compiler_f90,
                           posix=(os.name == 'posix'))
 
@@ -477,7 +484,7 @@ class FCompiler(CCompiler):
         fixflags = []
 
         if f77:
-            f77flags = self.flag_vars.f77
+            f77flags = f77_env_flags + self.flag_vars.f77
         if f90:
             f90flags = self.flag_vars.f90
             freeflags = self.flag_vars.free
@@ -517,7 +524,7 @@ class FCompiler(CCompiler):
         fflags = self.flag_vars.flags + dflags + oflags + aflags
 
         if f77:
-            self.set_commands(compiler_f77=f77+f77flags+fflags)
+            self.set_commands(compiler_f77=[f77]+f77flags+fflags)
         if f90:
             self.set_commands(compiler_f90=f90+freeflags+f90flags+fflags)
         if fix:
```

eric-wieser commented:
Looking some more, I think the new behavior is correct

If I run `CC="gcc --oops" make main` on linux, I get errors about unrecognized command line options, not about about a missing "gcc --oops" binary.

So I think it's reasonable to require executable names to be quoted if they contain spaces, both on windows and linux

charris commented:
In it goes. Thanks Tyler.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -136,20 +137,39 @@ def CCompiler_spawn(self, cmd, display=None):
+    except UnicodeError:
+        # When installing through pip, `o` can contain non-ascii chars
+        pass
+    if re.search(b'Too many open files', o):
```
I worry that this regex was intended to match stderr, in a similar way to #12831.
